File: apps/backend/src/main/java/org/mepcity/kursplatform/core/observability/SafeLogEvent.java
package org.mepcity.kursplatform.core.observability;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;

/** Provider-independent diagnostic event created only through typed factories. */
public final class SafeLogEvent {

	private static final Pattern REQUEST_ID = Pattern.compile("[A-Za-z0-9._:-]{1,128}");
	private static final Pattern ERROR_TYPE = Pattern.compile("[A-Za-z_$][A-Za-z0-9_.$]{0,127}");
	private static final Pattern STACK_PART = Pattern.compile("[A-Za-z_$][A-Za-z0-9_.$]{0,255}");
	private static final String APPLICATION_PACKAGE = "org.mepcity.kursplatform.";
	private static final Pattern ROUTE_PATTERN = Pattern.compile("/[A-Za-z0-9._~!$&'()*+,;=:@/{}`*\\[\\]-]{0,255}");

	private final String name;
	private final SafeLogSeverity severity;
	private final Instant occurredAt;
	private final Map<String, Object> fields;

	private SafeLogEvent(
			String name,
			SafeLogSeverity severity,
			Instant occurredAt,
			Map<String, Object> fields) {
		this.name = name;
		this.severity = Objects.requireNonNull(severity, "severity is required");
		this.occurredAt = Objects.requireNonNull(occurredAt, "occurredAt is required");
		this.fields = Map.copyOf(fields);
	}

	public static SafeLogEvent httpRequestCompleted(
			Instant occurredAt,
			String requestId,
			String method,
			String routePattern,
			int status,
			long durationMs) {
		return httpRequestCompleted(
				occurredAt, requestId, method, routePattern, status, durationMs, null);
	}

	public static SafeLogEvent httpRequestCompleted(
			Instant occurredAt,
			String requestId,
			String method,
			String routePattern,
			int status,
			long durationMs,
			RuntimeEnvironment environment) {
		validateRequestId(requestId);
		String validatedMethod = HttpMethod.fromValue(method).value();
		String validatedRoute = validateRoutePattern(routePattern);
		if (status < 100 || status > 599) {
			throw new IllegalArgumentException("Invalid HTTP status");
		}
		if (durationMs < 0) {
			throw new IllegalArgumentException("durationMs cannot be negative");
		}

		Map<String, Object> fields = new LinkedHashMap<>();
		fields.put("requestId", requestId);
		fields.put("method", validatedMethod);
		fields.put("path", validatedRoute);
		fields.put("status", status);
		fields.put("durationMs", durationMs);
		if (environment != null) {
			fields.put("environment", environment.value());
		}
		return new SafeLogEvent(
				"http.request.completed", severityForHttpStatus(status), occurredAt, fields);
	}

	public static SafeLogEvent unexpectedApplicationError(
			Instant occurredAt, String requestId, SafeLogOperation operation, Throwable error) {
		return applicationError(
				occurredAt, requestId, operation, error, SafeLogSeverity.ERROR);
	}

	/** Unexpected HTTP diagnostic restricted to root type and an application-owned frame. */
	public static SafeLogEvent unexpectedHttpDiagnostic(
			Instant occurredAt, String requestId, Throwable error) {
		validateRequestId(requestId);
		Throwable root = rootCause(Objects.requireNonNull(error, "error is required"));
		String errorType = diagnosticErrorType(root);
		return new SafeLogEvent(
				"application.error.unexpected",
				SafeLogSeverity.ERROR,
				occurredAt,
				Map.of(
						"requestId", requestId,
						"operation", SafeLogOperation.HTTP_REQUEST.value(),
						"errorType", errorType,
						"stackLocation", applicationStackLocation(root)));
	}

	public static SafeLogEvent fatalApplicationError(
			Instant occurredAt, String requestId, SafeLogOperation operation, Throwable error) {
		return applicationError(
				occurredAt, requestId, operation, error, SafeLogSeverity.FATAL);
	}

	/** Security operation events admit only pre-validated opaque identifiers, never payloads. */
	public static SafeLogEvent securityAlert(String name, SafeLogSeverity severity, Instant occurredAt,
			Map<String, String> attributes) {
		if (name == null || !name.matches("iam\\.security\\.[a-z_]{1,64}")) {
			throw new IllegalArgumentException("Invalid security alert name");
		}
		if (severity != SafeLogSeverity.WARNING && severity != SafeLogSeverity.ERROR) {
			throw new IllegalArgumentException("Invalid security alert severity");
		}
		if (attributes == null || attributes.entrySet().stream().anyMatch(entry ->
				!entry.getKey().matches("[a-zA-Z][a-zA-Z0-9]{0,63}")
						|| !entry.getValue().matches("[A-Za-z0-9._:-]{1,128}"))) {
			throw new IllegalArgumentException("Invalid security alert attributes");
		}
		return new SafeLogEvent(name, severity, occurredAt, new LinkedHashMap<>(attributes));
	}

	private static SafeLogEvent applicationError(
			Instant occurredAt,
			String requestId,
			SafeLogOperation operation,
			Throwable error,
			SafeLogSeverity severity) {
		validateRequestId(requestId);
		Objects.requireNonNull(operation, "operation is required");
		Objects.requireNonNull(error, "error is required");
		String errorType = validateErrorType(error.getClass().getSimpleName());
		return new SafeLogEvent("application.error.unexpected", severity, occurredAt, Map.of(
				"requestId", requestId,
				"operation", operation.value(),
				"errorType", errorType));
	}

	private static String validateErrorType(String errorType) {
		if (!ERROR_TYPE.matcher(errorType).matches()) {
			throw new IllegalArgumentException("Invalid errorType");
		}
		return errorType;
	}

	private static String diagnosticErrorType(Throwable error) {
		String errorType = error.getClass().getSimpleName();
		return ERROR_TYPE.matcher(errorType).matches() ? errorType : "Throwable";
	}

	private static String applicationStackLocation(Throwable error) {
		for (StackTraceElement frame : error.getStackTrace()) {
			if (frame.getClassName().startsWith(APPLICATION_PACKAGE)
					&& STACK_PART.matcher(frame.getClassName()).matches()
					&& STACK_PART.matcher(frame.getMethodName()).matches()) {
				return frame.getClassName() + "#" + frame.getMethodName() + ":"
						+ Math.max(0, frame.getLineNumber());
			}
		}
		return "unavailable";
	}

	private static Throwable rootCause(Throwable error) {
		Throwable current = error;
		for (int depth = 0;
				depth < 16 && current.getCause() != null && current.getCause() != current;
				depth++) {
			current = current.getCause();
		}
		return current;
	}

	private static SafeLogSeverity severityForHttpStatus(int status) {
		if (status >= 500) {
			return SafeLogSeverity.ERROR;
		}
		if (status >= 400) {
			return SafeLogSeverity.WARNING;
		}
		return SafeLogSeverity.INFO;
	}

	private static void validateRequestId(String requestId) {
		if (requestId == null || !REQUEST_ID.matcher(requestId).matches()) {
			throw new IllegalArgumentException("Invalid requestId");
		}
	}

	private static String validateRoutePattern(String routePattern) {
		if ("unmatched".equals(routePattern)) {
			return routePattern;
		}
		if (routePattern == null || !ROUTE_PATTERN.matcher(routePattern).matches()) {
			throw new IllegalArgumentException("Invalid route pattern");
		}
		return routePattern;
	}

	public String name() {
		return name;
	}

	public SafeLogSeverity severity() {
		return severity;
	}

	public Instant occurredAt() {
		return occurredAt;
	}

	public Map<String, Object> fields() {
		return fields;
	}

	public enum SafeLogOperation {
		HTTP_REQUEST("http.request");

		private final String value;

		SafeLogOperation(String value) {
			this.value = value;
		}

		public String value() {
			return value;
		}

		public static SafeLogOperation fromValue(String value) {
			for (SafeLogOperation operation : values()) {
				if (operation.value.equals(value)) {
					return operation;
				}
			}
			throw new IllegalArgumentException("Unsupported operation");
		}
	}

	public enum RuntimeEnvironment {
		DEVELOPMENT("development"),
		STAGING("staging"),
		PRODUCTION("production");

		private final String value;

		RuntimeEnvironment(String value) {
			this.value = value;
		}

		public String value() {
			return value;
		}

		public static RuntimeEnvironment fromValue(String value) {
			for (RuntimeEnvironment environment : values()) {
				if (environment.value.equals(value)) {
					return environment;
				}
			}
			throw new IllegalArgumentException("Unsupported environment");
		}
	}

	private enum HttpMethod {
		GET,
		HEAD,
		POST,
		PUT,
		PATCH,
		DELETE,
		OPTIONS,
		TRACE,
		CONNECT;

		private String value() {
			return name();
		}

		private static HttpMethod fromValue(String value) {
			try {
				return valueOf(value);
			} catch (NullPointerException | IllegalArgumentException exception) {
				throw new IllegalArgumentException("Invalid HTTP method", exception);
			}
		}
	}
}

File: apps/mobile/lib/core/observability/safe_log_event.dart
import 'dart:collection';

enum SafeLogSeverity { info, warning, error, fatal }

enum SafeLogOperation {
  httpRequest('http.request'),
  flutterFramework('flutter.framework'),
  flutterPlatform('flutter.platform'),
  dartZone('dart.zone');

  const SafeLogOperation(this.value);

  final String value;

  static SafeLogOperation parse(String value) {
    for (final operation in values) {
      if (operation.value == value) {
        return operation;
      }
    }
    throw ArgumentError.value(value, 'value', 'Unsupported operation');
  }
}

enum SafeRuntimeEnvironment {
  development('development'),
  staging('staging'),
  production('production');

  const SafeRuntimeEnvironment(this.value);

  final String value;

  static SafeRuntimeEnvironment parse(String value) {
    for (final environment in values) {
      if (environment.value == value) {
        return environment;
      }
    }
    throw ArgumentError.value(value, 'value', 'Unsupported environment');
  }
}

/// Provider-independent diagnostic event created only through typed factories.
final class SafeLogEvent {
  SafeLogEvent._({
    required this.name,
    required this.severity,
    required this.occurredAt,
    required Map<String, Object> fields,
  }) : _fields = Map.unmodifiable(fields);

  static final _requestId = RegExp(r'^[A-Za-z0-9._:-]{1,128}$');
  static final _errorType = RegExp(r'^[A-Za-z_$][A-Za-z0-9_.$]{0,127}$');
  static final _routePattern = RegExp(
    r"^/[A-Za-z0-9._~!$&'()*+,;=:@/{}`*\[\]-]{0,255}$",
  );
  static const _httpMethods = <String>{
    'GET',
    'HEAD',
    'POST',
    'PUT',
    'PATCH',
    'DELETE',
    'OPTIONS',
    'TRACE',
    'CONNECT',
  };

  final String name;
  final SafeLogSeverity severity;
  final DateTime occurredAt;
  final Map<String, Object> _fields;

  factory SafeLogEvent.httpRequestCompleted({
    required DateTime occurredAt,
    required String requestId,
    required String method,
    required String routePattern,
    required int status,
    required int durationMs,
    SafeRuntimeEnvironment? environment,
  }) {
    _validateRequestId(requestId);
    if (!_httpMethods.contains(method)) {
      throw ArgumentError.value(method, 'method', 'Invalid HTTP method');
    }
    if (routePattern != 'unmatched' && !_routePattern.hasMatch(routePattern)) {
      throw ArgumentError.value(
        routePattern,
        'routePattern',
        'Invalid route pattern',
      );
    }
    if (status < 100 || status > 599) {
      throw ArgumentError.value(status, 'status', 'Invalid HTTP status');
    }
    if (durationMs < 0) {
      throw ArgumentError.value(
        durationMs,
        'durationMs',
        'durationMs cannot be negative',
      );
    }
    return SafeLogEvent._(
      name: 'http.request.completed',
      severity: _severityForStatus(status),
      occurredAt: occurredAt,
      fields: <String, Object>{
        'requestId': requestId,
        'method': method,
        'path': routePattern,
        'status': status,
        'durationMs': durationMs,
        if (environment != null) 'environment': environment.value,
      },
    );
  }

  factory SafeLogEvent.unexpectedApplicationError({
    required DateTime occurredAt,
    required SafeLogOperation operation,
    required Object error,
    String? requestId,
    SafeRuntimeEnvironment? environment,
    SafeLogSeverity severity = SafeLogSeverity.error,
  }) {
    if (requestId != null) {
      _validateRequestId(requestId);
    }
    if (severity != SafeLogSeverity.error &&
        severity != SafeLogSeverity.fatal) {
      throw ArgumentError.value(
        severity,
        'severity',
        'Unexpected errors require error or fatal severity',
      );
    }
    final errorType = error.runtimeType.toString();
    validateErrorType(errorType);
    final environmentValue = environment?.value;
    return SafeLogEvent._(
      name: 'application.error.unexpected',
      severity: severity,
      occurredAt: occurredAt,
      fields: <String, Object>{
        'operation': operation.value,
        'errorType': errorType,
        'requestId': ?requestId,
        'environment': ?environmentValue,
      },
    );
  }

  Map<String, Object> toJson() => UnmodifiableMapView(<String, Object>{
    'event': name,
    'severity': severity.name,
    'occurredAt': occurredAt.toUtc().toIso8601String(),
    ..._fields,
  });

  static SafeLogSeverity _severityForStatus(int status) {
    if (status >= 500) {
      return SafeLogSeverity.error;
    }
    if (status >= 400) {
      return SafeLogSeverity.warning;
    }
    return SafeLogSeverity.info;
  }

  static void _validateRequestId(String requestId) {
    if (!_requestId.hasMatch(requestId)) {
      throw ArgumentError.value(requestId, 'requestId', 'Invalid requestId');
    }
  }

  /// Exposed for negative contract tests; event creation still requires a typed factory.
  static void validateErrorType(String errorType) {
    if (!_errorType.hasMatch(errorType)) {
      throw ArgumentError.value(errorType, 'errorType', 'Invalid errorType');
    }
  }
}
